### PR TITLE
[Feature](mluOpCore): access variable in tensor struct through function in tensor.cpp and tensor.h

### PR DESCRIPTION
**Pull request by nth-BYTE**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

Please describe your motivation and the goal you want to achieve through this pull request.

## 2. Modification

In mluOpTensorStruct , pubilc variables have been changed into private  variables, except those will be removed in future， e.g. position,positions,scale,scales,offset,offsets. And then, some unused methods are deleted ,including tensorDimN,tensorDimC,tensorDimH,tensorDimW

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS™ Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [ ] MLU370
  - [ ] MLU590
- Job types
  - [ ] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS™-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS™ Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
# The test results should contain Op name, Shape, Data type,  
#   MLU Hardware Time(us), MLU Interface Time(us), MLU IO Efficiency, 
#   MLU Compute Efficiency, and Mlu Workspace Size(Bytes)
# 
# for example:
#
# ----------- case0 -----------
# case0
# [Op name                ]: abs
# [Shape                  ]: input.shape=[1024,1024,3,4], output.shape=[1024,1024,3,4]
# [Data type]             ]: float32
# [MLU Hardware Time      ]: 15728 (us)
# [MLU Interface Time     ]: 369.008 (us)
# [MLU IO Efficiency      ]: 0.23275
# [MLU Compute Efficiency ]: 0.5
# [Mlu Workspace Size     ]: -1 (Bytes)
# 
# ----------- case1 -----------
# ...
```

Platform：MLU590
```bash
# ----------- case0 -----------
# ----------- case1 -----------
# ...
```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: core/tensor.h [hunk 1]
```diff
@@ -107,6 +107,152 @@ struct alignas(64) mluOpTensorStruct {
...
+      const int *group_dimSize, const int desc_num);
+
+  mluOpStatus_t setGroupTensorDescriptors_v2(
+      mluOpTensorDescriptor_t **group_desc,
+      const mluOpTensorLayout_t *group_layout,
+      const mluOpDataType_t *group_dtype, const int *group_dimNb,
+      const int64_t *group_dimSize, const int desc_num);
+
+  uint64_t getTensorElementNum() { return this->total_element_num; }
+  // private:
```
Comment (zhouyue624) [inline core/tensor.h:255]: position/scale/offset/positions/scales/offsets 这几个跟量化相关的成员变量计划从 TensorStruct 摘出去，所以先保持现状。

File: core/tensor.h [hunk 2]
```diff
@@ -107,6 +107,152 @@ struct alignas(64) mluOpTensorStruct {
...
+  }
+
+  inline uint64_t getTotalTensorSize() const { return this->total_tensor_size; }
+  inline void setTotalTensorSize(uint64_t newSize) {
+    this->total_tensor_size = newSize;
+  }
+  inline uint64_t getTotalElementNum() const { return this->total_element_num; }
+  inline void setTotalElementNum(uint64_t newNum) {
+    this->total_element_num = newNum;
+  }
```
Comment (zhouyue624) [inline core/tensor.h:129]: total_tensor_size 和 total_element_num 是只读的，只提供 getter，它们的值是在其它数据成员的 setter 中联动设置的。

File: core/tensor.h [hunk 3]
```diff
@@ -107,6 +107,152 @@ struct alignas(64) mluOpTensorStruct {
...
+
+  inline int getDim() const { return this->dim; }
+  inline void setDim(int newDim) { this->dim = newDim; }
+  inline void releaseDims() { delete[] this->dims; }
+  inline int64_t *getDims() const { return this->dims; }
+  inline int64_t getDimIndex(size_t index) { return (this->dims)[index]; }
+  inline void setDims(int64_t *newDims) { this->dims = newDims; }
+
+  inline void releaseStrides() { delete[] this->strides; }
+  inline int64_t *getStrides() const { return this->strides; }
```
Comment (zhouyue624) [inline core/tensor.h:149]: releaseXxx 没必要也不应该单独包装成方法，如果编码中出现了这样的需求，那就应该先检讨下设计思路。

File: core/tensor.h [hunk 4]
```diff
@@ -107,6 +107,152 @@ struct alignas(64) mluOpTensorStruct {
...
+      mluOpTensorDescriptor_t **group_desc,
+      const mluOpTensorLayout_t *group_layout,
+      const mluOpDataType_t *group_dtype, const int *group_dimNb,
+      const int *group_dimSize, const int desc_num);
+
+  mluOpStatus_t setGroupTensorDescriptors_v2(
+      mluOpTensorDescriptor_t **group_desc,
+      const mluOpTensorLayout_t *group_layout,
+      const mluOpDataType_t *group_dtype, const int *group_dimNb,
+      const int64_t *group_dimSize, const int desc_num);
```
Comment (zhouyue624) [inline core/tensor.h:252]: GroupTensorDescriptor 也类似，不放在它所管理的对象的内部。

File: core/tensor.h (hunk 3 above)
Comment (nth-BYTE) [reply core/tensor.h:149]: release是上个版本修改的时候加的，忘记还原这个了，已修改

File: core/tensor.h [hunk 5]
```diff
@@ -107,6 +107,152 @@ struct alignas(64) mluOpTensorStruct {
...
+                                                          float *scale,
+                                                          int *offset);
+
+  mluOpStatus_t getTensorDescriptorPointerMode(
+      mluOpPointerMode_t *pointer_mode);
+
+  mluOpStatus_t destroyTensorDescriptor();
+
+  mluOpStatus_t destroyGroupTensorDescriptors(
+      mluOpTensorDescriptor_t **group_desc, const int desc_num);
```
[the remark this answers is not in the document]
Comment (nth-BYTE) [reply core/tensor.h:238]: 已去除

File: core/tensor.h (hunk 4 above)
Comment (nth-BYTE) [reply core/tensor.h:252]: 已去除

File: core/tensor.h (hunk 1 above)
Comment (nth-BYTE) [reply core/tensor.h:234]: 已去除

File: core/tensor.h [hunk 6]
```diff
@@ -288,16 +393,16 @@ inline int mluOpDataTypeBytes(const mluOpDataType_t dt) {
...
 
 inline int64_t mluOpGetTensordimN(const mluOpTensorDescriptor_t desc) {
-  switch (desc->layout) {
+  switch (desc->getLayout()) {
     case MLUOP_LAYOUT_NCHW:
     case MLUOP_LAYOUT_NHWC:
     case MLUOP_LAYOUT_NDHWC:
     case MLUOP_LAYOUT_NLC:
-      return desc->dims[0];
+      return desc->getDimIndex(0);
```
Comment (jerryjiahaha) [inline core/tensor.h:401]: 这里和下面都是dim[0]，可以删除这一行

File: core/tensor.h [hunk 7]
```diff
@@ -319,18 +424,18 @@ inline int64_t mluOpGetTensordimD(const mluOpTensorDescriptor_t desc) {
...
     case MLUOP_LAYOUT_NDHWC:
-      return desc->dims[4];
+      return desc->getDimIndex(4);
     case MLUOP_LAYOUT_NCDHW:
-      return desc->dims[1];
+      return desc->getDimIndex(1);
     case MLUOP_LAYOUT_HWCN:
     case MLUOP_LAYOUT_NLC:
-      return desc->dims[2];
+      return desc->getDimIndex(2);
```
Comment (jerryjiahaha) [inline core/tensor.h:438]: 这里要么layout和 mlu_op.h 里的枚举值大小顺序保持一直，要么根据实际的index值做好归类。目前的排列顺序两种都不沾边。

以及有些表达Channel含义的新增枚举这里也没涉及，可以一并更新上？

Comment (jerryjiahaha) [reply core/tensor.h:438]: 后面几个同理

File: core/tensor.h [hunk 8]
```diff
@@ -466,12 +571,12 @@ inline int64_t mluOpGetSeqDataDimC(const mluOpSeqDataDescriptor_t desc) {
 
 inline uint64_t shapeStrideCount(const mluOpTensorDescriptor_t desc) {
```
Comment (jerryjiahaha) [inline core/tensor.h:572]: 这个函数为啥不直接返回 `total_element_num` ？可以看下 tensor.cpp中`mluOpTensorStruct::setTensorDescriptorEx` 里面的逻辑。

并且这么一想，`setDim` 这个接口暴露出来也不太合适，应该直接setDims，然后setTensorDescriptorEx里调用它

File: core/tensor.h (hunk 2 above)
Comment (nth-BYTE) [reply core/tensor.h]: 已去除

File: core/tensor.h [hunk 9]
```diff
@@ -107,6 +107,111 @@ struct alignas(64) mluOpTensorStruct {
...
+  inline mluOpDataType_t getDtype() const { return this->dtype; }
+  inline void setDtype(mluOpDataType_t newDtype) { this->dtype = newDtype; }
+  inline mluOpDataType_t getOnchipDtype() const { return this->onchip_dtype; }
+  inline void setOnchipDtype(mluOpDataType_t newDtype) {
+    this->onchip_dtype = newDtype;
+  }
+
+  inline int getDim() const { return this->dim; }
+  inline void setDim(int newDim) { this->dim = newDim; }
+  inline int64_t *getDims() const { return this->dims; }
```
[the remark this answers is not in the document]
Comment (nth-BYTE) [reply core/tensor.h:129]: getDims已修改。
整体结构留待调整，并且出于性能考虑会用别的设计，具体另起jira。

File: core/tensor.h (hunk 8 above)
Comment (nth-BYTE) [reply core/tensor.h:572]: 这边的值不是total_element_num,setDim已修改，改为接口较用setDims

File: core/tensor.h (hunk 7 above)
Comment (nth-BYTE) [reply core/tensor.h:435]: 已修改

Comment (jerryjiahaha) [reply core/tensor.h:437]: 仔细想了下，给这种接口感觉意义还是不够方便，每个维度都要做调用，每次调用内部还switch-case。不如再搞个接口直接返回多个值？比如
```
auto [n,h,w,c] = getTensorDimNHWC(desc)  // 对于nhwc/nchw这些包含必要信息的生效。
```

File: core/tensor.h (hunk 8 above)
Comment (jerryjiahaha) [reply core/tensor.h:572]: total_element_num确实是我搞错了。setDims没找到在哪里。

shapeStrideCount这个命名有点奇怪。torch里面是Storage类的size方法返回同样的信息。我个人感觉叫做storageElementNum或者storageSize这种会好点。

还有这个方法为什么不作为Tensor struct的成员方法，而要这么单独写？

Comment (nth-BYTE) [reply core/tensor.h:572]: shapeStrideCount，我后面去跟这个修改的提交人对一下。

File: core/tensor.cpp [hunk 10]
```diff
@@ -40,6 +40,26 @@
...
+    return MLUOP_STATUS_BAD_PARAM;
+  }
+}
+
+static inline void mluOpSetTensorDescriptorDimBase(mluOpTensorDescriptor_t desc,
+                                                   int dimNb) {
+  desc->setTensorDescriptorDimBase(dimNb);                                             
+}
 
 mluOpStatus_t mluOpTensorStruct::tensorDimN(size_t &dim_out) {
```
[the remark this answers is not in the document]
Comment (nth-BYTE) [reply core/tensor.cpp:59]: 后面跟写这些接口的人员对接一下，确认下添加的意图，如有修改必要则另起jira。

File: core/tensor.h (hunk 8 above)
Comment (nth-BYTE) [reply core/tensor.h:572]: setDims之前版本用到了，现在确实没用，删除掉了

**Review by jerryjiahaha — APPROVED** [review]

**Review by zhouyue624 — APPROVED** [review]

**Review by PetrelYy — APPROVED** [review]
LGTM

**Pull request merged**